cf-agent/package_module.c

```diff
@@ -0,0 +1,1604 @@
+    free(wrapper->path);
+    free(wrapper->name);
+    free(wrapper);
+}
+
+PackageModuleWrapper *NewPackageModuleWrapper(PackageModuleBody *package_module)
+{
+    assert(package_module && package_module->name);
+
+    PackageModuleWrapper *wrapper = xmalloc(sizeof(PackageModuleWrapper));
+    if (!wrapper)
+    {
+        return NULL;
+    }
+    
+    wrapper->path = GetPackageModuleRealPath(package_module->name);
+    wrapper->name = SafeStringDuplicate(package_module->name);
+    wrapper->package_module = package_module;
+    
+    /* Check if file exists */
+    struct stat sb;
+    if (!wrapper->path || (stat(wrapper->path, &sb) != 0))
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "can not find package wrapper in provided location: %s",
+            wrapper->path);
+        DeletePackageModuleWrapper(wrapper);
+        return NULL;
+    }
+    
+    /* Negotiate API version */
+    wrapper->supported_api_version = NegotiateSupportedAPIVersion(wrapper);
+    if (wrapper->supported_api_version != 1)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "unsupported package module wrapper API version: %d",
+            wrapper->supported_api_version);
+        DeletePackageModuleWrapper(wrapper);
+        return NULL;
+    }
+    return wrapper;
+}
+
+
+void UpdatePackagesCache(EvalContext *ctx, bool force_update)
+{
+    Log(LOG_LEVEL_DEBUG, "Updating package cache.");
+    
+    PackagePromoseGlobalLock package_lock = 
+            AcquireGlobalPackagePromiseLock(ctx);
+                        
+    if (package_lock.g_lock.lock == NULL)
+    {
+        Log(LOG_LEVEL_INFO, "Can not aquire global lock for package promise. "
+            "Skipping updating cache.");
+        return;
+    }
+                        
+    Rlist *default_inventory = GetDefaultInventoryFromContext(ctx);
+    
+    for (const Rlist *rp = default_inventory; rp != NULL; rp = rp->next)
+    {
+        const char *pm_name =  RlistScalarValue(rp);
+        
+        /* We don't want inventory to be reported. */
+        if (StringSafeEqual(pm_name, "cf_null"))
+        {
+            break;
+        }
+        
+        PackageModuleBody *module = GetPackageModuleFromContext(ctx, pm_name);
+        if (!module)
+        {
+            Log(LOG_LEVEL_VERBOSE,
+                "Can not find body for package module: %s", pm_name);
+            continue;
+        }
+
+        PackageModuleWrapper *module_wrapper = NewPackageModuleWrapper(module);
+
+        if (!module_wrapper)
+        {
+            Log(LOG_LEVEL_VERBOSE,
+                "Can not set up wrapper for module: %s", pm_name);
+            continue;
+        }
+
+        UpdateSinglePackageModuleCache(ctx, module_wrapper,
+                                       UPDATE_TYPE_INSTALLED, force_update);
+        UpdateSinglePackageModuleCache(ctx, module_wrapper,
+                                       force_update ? UPDATE_TYPE_LOCAL_UPDATES : 
+                                           UPDATE_TYPE_UPDATES,
```

Comment:
Is this correct? My understanding is that online updates should *never* be forced, not even after installing a package. What's important is to update CFEngine cache to include the updates that apt and yum know about. Going online is of less importance, and if you install several packages after one another, is close to useless, since nothing will have changed since the previous online query.